internal/functions/efgs/download-batch.go

```diff
@@ -161,9 +201,19 @@ func downloadAllRecursively(ctx context.Context, config *downloadConfig, now tim
 		// This ends once
 		if moreKeysAvailable {
 			keys = append(keys, moreKeys...)
-			params = nextBatchParams(ctx, now, params)
+			nextBatch = nextBatchParams(ctx, now, nextBatch)
 		} else {
-			logger.Infof("Batch %v doesn't exist, stopping", params.BatchTag)
+			logger.Infof("Batch %v doesn't exist, stopping", nextBatch.BatchTag)
+			break
+		}
+
+		if len(keys) >= config.MaxDownloadYesterdaysKeysPartSize {
+			// There's too much of keys; let's process only part and prevent timeout.
+			logger.Infof("There's too much of keys, about to postpone processing of the rest")
+			batchToPostpone = &nextBatch
+
+			// now break the downloading and enqueue what we've got so far
+			break
```

Comment:
It'd make more sense to me to really split downloading and processing. Let this function only download stuff and don't process anything at all, even if the batch is very small. Just download & enqueue and let "something else" to handle the rest. Wdyt?

Reply:
Well, "processing" means that keys are imported to the keys server and that already **is** done somewhere else (`EfgsImportKeys` function). What this function does is downloading keys, sorting them into batches, and enqueue the batches for processing by the _import_ function. Originally, the download and sorting were done here at once because of batches size optimization but, I have to admit, it just needs a big-enough input set and not an "all items" input set.  
I have one more idea how to make it maybe more logical - change the code so everything would be done in _Postpone_ function and the one thing that would `EfgsDownloadYesterdaysKeys` do is that it would trigger the initial _Postpone_ :-)